File: src/benchmark.cpp
#include "regjit.h"
#include <iostream>
#include <chrono>
#include <regex>
#include <vector>
#include <string>
#include <iomanip>

// Benchmark result structure
struct BenchResult {
    std::string name;
    std::string pattern;
    std::string input;
    uint64_t jit_ns;
    uint64_t std_ns;
    double speedup;
};

// Number of iterations for each benchmark
constexpr int ITERATIONS = 100000;
constexpr int WARMUP = 1000;

// Benchmark JIT regex
uint64_t benchmark_jit(const std::string& pattern, const std::string& input) {
    Initialize();
    if (!CompileRegex(pattern)) {
        std::cerr << "Failed to compile pattern: " << pattern << std::endl;
        CleanUp();
        return 0;
    }
    
    // Use the FunctionName set by CompileRegex
    std::string lookupName = FunctionName.empty() ? "match" : FunctionName;
    auto MatchSym = ExitOnErr(JIT->lookup(lookupName));
    auto MatchFunc = (int (*)(const char*))MatchSym.getValue();
    
    // Warmup
    for (int i = 0; i < WARMUP; ++i) {
        MatchFunc(input.c_str());
    }
    
    // Benchmark
    auto start = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < ITERATIONS; ++i) {
        MatchFunc(input.c_str());
    }
    auto end = std::chrono::high_resolution_clock::now();
    
    CleanUp();
    return std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count() / ITERATIONS;
}

// Benchmark std::regex
uint64_t benchmark_std_regex(const std::string& pattern, const std::string& input) {
    std::regex re(pattern);
    
    // Warmup
    for (int i = 0; i < WARMUP; ++i) {
        std::regex_search(input, re);
    }
    
    // Benchmark
    auto start = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < ITERATIONS; ++i) {
        std::regex_search(input, re);
    }
    auto end = std::chrono::high_resolution_clock::now();
    
    return std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count() / ITERATIONS;
}

// Run a single benchmark case
BenchResult run_benchmark(const std::string& name, const std::string& pattern, 
                          const std::string& input, const std::string& std_pattern = "") {
    BenchResult result;
    result.name = name;
    result.pattern = pattern;
    result.input = input.length() > 20 ? input.substr(0, 17) + "..." : input;
    
    result.jit_ns = benchmark_jit(pattern, input);
    
    // Use std_pattern if provided (for patterns that differ between JIT and std::regex)
    const std::string& p = std_pattern.empty() ? pattern : std_pattern;
    result.std_ns = benchmark_std_regex(p, input);
    
    result.speedup = result.jit_ns > 0 ? (double)result.std_ns / result.jit_ns : 0;
    
    return result;
}

void print_results(const std::vector<BenchResult>& results) {
    std::cout << "\n";
    std::cout << std::string(100, '=') << "\n";
    std::cout << "                           RegJIT vs std::regex Benchmark Results\n";
    std::cout << std::string(100, '=') << "\n";
    std::cout << std::left << std::setw(25) << "Test Case"
              << std::setw(20) << "Pattern"
              << std::right << std::setw(12) << "JIT (ns)"
              << std::setw(15) << "std::regex"
              << std::setw(12) << "Speedup"
              << "\n";
    std::cout << std::string(100, '-') << "\n";
    
    double total_speedup = 0;
    int count = 0;
    
    for (const auto& r : results) {
        std::cout << std::left << std::setw(25) << r.name
                  << std::setw(20) << r.pattern
                  << std::right << std::setw(12) << r.jit_ns
                  << std::setw(15) << r.std_ns
                  << std::setw(10) << std::fixed << std::setprecision(1) << r.speedup << "x"
                  << "\n";
        if (r.speedup > 0) {
            total_speedup += r.speedup;
            count++;
        }
    }
    
    std::cout << std::string(100, '-') << "\n";
    std::cout << std::left << std::setw(25) << "Average Speedup:"
              << std::right << std::setw(67) << std::fixed << std::setprecision(1) 
              << (count > 0 ? total_speedup / count : 0) << "x\n";
    std::cout << std::string(100, '=') << "\n";
}

int main() {
    std::vector<BenchResult> results;
    
    std::cout << "Running RegJIT vs std::regex benchmarks...\n";
    std::cout << "Iterations per test: " << ITERATIONS << "\n";
    
    // ========== Basic Character Matching ==========
    std::cout << "\n[1/8] Basic character matching..." << std::flush;
    results.push_back(run_benchmark(
        "Simple literal",
        "hello",
        "hello world"
    ));
    std::cout << " done\n";
    
    std::cout << "[2/8] Long literal matching..." << std::flush;
    results.push_back(run_benchmark(
        "Long literal",
        "abcdefghij",
        "xxxxxxxxxxabcdefghijyyyyyyyyyy"
    ));
    std::cout << " done\n";
    
    // ========== Quantifiers ==========
    std::cout << "[3/8] Quantifier benchmarks..." << std::flush;
    
    // a{1000} matching 1000 'a's
    std::string long_a(1000, 'a');
    results.push_back(run_benchmark(
        "Exact repeat {1000}",
        "a{1000}",
        long_a
    ));
    
    // a+ matching many 'a's
    results.push_back(run_benchmark(
        "Plus quantifier a+",
        "a+",
        long_a
    ));
    
    // a* matching many 'a's
    results.push_back(run_benchmark(
        "Star quantifier a*",
        "a*",
        long_a
    ));
    std::cout << " done\n";
    
    // ========== Character Classes ==========
    std::cout << "[4/8] Character class benchmarks..." << std::flush;
    
    std::string alphanum = "abc123XYZ789def456GHI";
    results.push_back(run_benchmark(
        "Char class [a-z]+",
        "[a-z]+",
        alphanum
    ));
    
    results.push_back(run_benchmark(
        "Char class [a-zA-Z0-9]+",
        "[a-zA-Z0-9]+",
        alphanum
    ));
    
    results.push_back(run_benchmark(
        "Negated class [^0-9]+",
        "[^0-9]+",
        alphanum
    ));
    std::cout << " done\n";
    
    // ========== Escape Sequences ==========
    std::cout << "[5/8] Escape sequence benchmarks..." << std::flush;
    
    std::string digits = "1234567890";
    std::string words = "hello_world_123";
    std::string mixed = "  \t\n  text  \r\n  ";
    
    results.push_back(run_benchmark(
        "Digit \\d+",
        "\\d+",
        digits
    ));
    
    results.push_back(run_benchmark(
        "Word \\w+",
        "\\w+",
        words
    ));
    
    results.push_back(run_benchmark(
        "Whitespace \\s+",
        "\\s+",
        mixed
    ));
    std::cout << " done\n";
    
    // ========== Alternation ==========
    std::cout << "[6/8] Alternation benchmarks..." << std::flush;
    
    results.push_back(run_benchmark(
        "Simple alternation",
        "cat|dog|bird",
        "I have a dog"
    ));
    
    results.push_back(run_benchmark(
        "Complex alternation",
        "hello|world|foo|bar|baz",
        "the world is beautiful"
    ));
    std::cout << " done\n";
    
    // ========== Anchors ==========
    std::cout << "[7/8] Anchor benchmarks..." << std::flush;
    
    results.push_back(run_benchmark(
        "Start anchor ^hello",
        "^hello",
        "hello world"
    ));
    
    results.push_back(run_benchmark(
        "End anchor world$",
        "world$",
        "hello world"
    ));
    
    results.push_back(run_benchmark(
        "Both anchors ^...$",
        "^hello world$",
        "hello world"
    ));
    std::cout << " done\n";
    
    // ========== Complex Patterns ==========
    std::cout << "[8/8] Complex pattern benchmarks..." << std::flush;
    
    // Email-like pattern
    results.push_back(run_benchmark(
        "Email-like pattern",
        "[a-z]+@[a-z]+\\.[a-z]+",
        "contact user@example.com for info"
    ));
    
    // IP-like pattern (simplified)
    results.push_back(run_benchmark(
        "IP-like pattern",
        "\\d+\\.\\d+\\.\\d+\\.\\d+",
        "Server IP is 192.168.1.100"
    ));
    
    // Nested groups
    results.push_back(run_benchmark(
        "Nested groups",
        "(a(b(c)+)+)+",
        "abcbcabcbcbc"
    ));
    
    // Long input search
    std::string long_text(10000, 'x');
    long_text += "needle";
    results.push_back(run_benchmark(
        "Long input search",
        "needle",
        long_text
    ));
    std::cout << " done\n";
    
    // Print results
    print_results(results);
    
    return 0;
}

File: python/benchmark.py
#!/usr/bin/env python3
"""
RegJIT vs Python re Benchmark

Compares RegJIT (via Python bindings) against Python's standard `re` module.
This benchmark mirrors the C++ benchmark.cpp for Python re compatibility.
"""

import sys
import os
import timeit
import re
from typing import Dict, List, Tuple, Optional

# Add python directory to path for _regjit module
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

# Benchmark settings (matching C++ benchmark.cpp)
ITERATIONS = 10000
WARMUP = 100


class BenchResult:
    def __init__(self, name: str, pattern: str, input_str: str):
        self.name = name
        self.pattern = pattern
        self.input = input_str[:17] + "..." if len(input_str) > 20 else input_str
        self.jit_ns = 0
        self.re_ns = 0
        self.speedup_vs_re = 0.0


def benchmark_re(
    pattern: str, input_str: str, iterations: int = ITERATIONS, warmup: int = WARMUP
) -> float:
    """Benchmark Python re module."""
    # Compile pattern
    compiled = re.compile(pattern)

    # Warmup
    for _ in range(warmup):
        compiled.search(input_str)

    # Benchmark
    start = timeit.default_timer()
    for _ in range(iterations):
        compiled.search(input_str)
    end = timeit.default_timer()

    # Return nanoseconds per iteration
    return (end - start) * 1_000_000_000 / iterations


def benchmark_jit(
    pattern: str, input_str: str, iterations: int = ITERATIONS, warmup: int = WARMUP
) -> float:
    """Benchmark RegJIT via Python bindings."""
    from _regjit import Regex

    # Compile pattern
    jit_regex = Regex(pattern)

    # Warmup
    for _ in range(warmup):
        jit_regex.match(input_str)

    # Benchmark
    start = timeit.default_timer()
    for _ in range(iterations):
        jit_regex.match(input_str)
    end = timeit.default_timer()

    # Return nanoseconds per iteration
    return (end - start) * 1_000_000_000 / iterations


def run_benchmark(name: str, pattern: str, input_str: str) -> BenchResult:
    """Run a single benchmark case."""
    result = BenchResult(name, pattern, input_str)

    print(f"  Testing: {name} ({pattern})...", end=" ", flush=True)

    # Benchmark RegJIT
    try:
        result.jit_ns = benchmark_jit(pattern, input_str)
    except Exception as e:
        print(f"\n  ERROR: RegJIT failed for pattern '{pattern}': {e}")
        result.jit_ns = 0

    # Benchmark Python re
    try:
        result.re_ns = benchmark_re(pattern, input_str)
    except Exception as e:
        print(f"\n  ERROR: Python re failed for pattern '{pattern}': {e}")
        result.re_ns = 0

    # Calculate speedup
    if result.jit_ns > 0:
        result.speedup_vs_re = result.re_ns / result.jit_ns

    print("done")

    return result


def print_results(results: List[BenchResult]):
    """Print benchmark results in a table format."""
    print("\n" + "=" * 100)
    print("                        RegJIT vs Python re Benchmark Results")
    print("=" * 100)
    print(
        f"{'Test Case':<22} {'Pattern':<18} {'RegJIT(ns)':<12} {'Python re':<14} {'Speedup':<10}"
    )
    print("-" * 100)

    total_speedup = 0.0
    count = 0

    for r in results:
        pattern_display = r.pattern[:13] + "..." if len(r.pattern) > 16 else r.pattern
        speedup_display = f"{r.speedup_vs_re:.2f}x" if r.speedup_vs_re > 0 else "N/A"

        print(
            f"{r.name:<22} {pattern_display:<18} {r.jit_ns:<12.0f} {r.re_ns:<14.0f} {speedup_display:<10}"
        )

        if r.speedup_vs_re > 0:
            total_speedup += r.speedup_vs_re
            count += 1

    print("-" * 100)

    avg_speedup = total_speedup / count if count > 0 else 0
    print(f"{'Average Speedup:':<42} {'':<12} {'':<14} {avg_speedup:.2f}x")
    print("=" * 100)


def main():
    print("Running RegJIT vs Python re benchmarks...")
    print(f"Iterations per test: {ITERATIONS}")
    print(f"Warmup iterations: {WARMUP}")

    results: List[BenchResult] = []

    # ========== Basic Character Matching ==========
    print("\n[1/8] Basic character matching...")
    results.append(run_benchmark("Simple literal", "hello", "hello world"))
    results.append(
        run_benchmark("Long literal", "abcdefghij", "xxxxxxxxxxabcdefghijyyyyyyyyyy")
    )

    # ========== Quantifiers ==========
    print("\n[2/8] Quantifier benchmarks...")
    long_a = "a" * 1000
    results.append(run_benchmark("Exact repeat {1000}", "a{1000}", long_a))
    results.append(run_benchmark("Plus quantifier a+", "a+", long_a))
    results.append(run_benchmark("Star quantifier a*", "a*", long_a))

    # ========== Character Classes ==========
    print("\n[3/8] Character class benchmarks...")
    alphanum = "abc123XYZ789def456GHI"
    results.append(run_benchmark("Char class [a-z]+", "[a-z]+", alphanum))
    results.append(run_benchmark("Char class [a-zA-Z0-9]+", "[a-zA-Z0-9]+", alphanum))
    results.append(run_benchmark("Negated class [^0-9]+", "[^0-9]+", alphanum))

    # ========== Escape Sequences ==========
    print("\n[4/8] Escape sequence benchmarks...")
    digits = "1234567890"
    words = "hello_world_123"
    mixed = "  \t\n  text  \r\n  "
    results.append(run_benchmark("Digit \\d+", r"\d+", digits))
    results.append(run_benchmark("Word \\w+", r"\w+", words))
    results.append(run_benchmark("Whitespace \\s+", r"\s+", mixed))

    # ========== Alternation ==========
    print("\n[5/8] Alternation benchmarks...")
    results.append(run_benchmark("Simple alternation", "cat|dog|bird", "I have a dog"))
    results.append(
        run_benchmark(
            "Complex alternation", "hello|world|foo|bar|baz", "the world is beautiful"
        )
    )

    # ========== Anchors ==========
    print("\n[6/8] Anchor benchmarks...")
    results.append(run_benchmark("Start anchor ^hello", "^hello", "hello world"))
    results.append(run_benchmark("End anchor world$", "world$", "hello world"))
    results.append(run_benchmark("Both anchors ^...$", "^hello world$", "hello world"))

    # ========== Complex Patterns ==========
    print("\n[7/8] Complex pattern benchmarks...")
    results.append(
        run_benchmark(
            "Email-like pattern",
            r"[a-z]+@[a-z]+\.[a-z]+",
            "contact user@example.com for info",
        )
    )
    results.append(
        run_benchmark(
            "IP-like pattern", r"\d+\.\d+\.\d+\.\d+", "Server IP is 192.168.1.100"
        )
    )
    results.append(run_benchmark("Nested groups", r"(a(b(c)+)+)+", "abcbcabcbcbc"))

    # Long input search
    print("\n[8/8] Long input search...")
    long_text = "x" * 10000 + "needle"
    results.append(run_benchmark("Long input search", "needle", long_text))

    # Print results
    print_results(results)

    return 0


if __name__ == "__main__":
    sys.exit(main())
